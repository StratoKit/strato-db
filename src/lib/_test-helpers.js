import DB from '../DB'
import ESDB from '../EventSourcingDB'
import EQ from '../EventQueue'
import JsonModel from '../JsonModel'

export {DB, JsonModel}

export const getModel = options => {
	const db = new DB()
	return db.addModel(JsonModel, {
		name: 'testing',
		keepRowId: false,
		...options,
	})
}

export const sharedSetup = getPromise => fn => {
	let promise
	return async () => {
		if (!promise) {
			promise = getPromise()
		}
		return fn(await promise)
	}
}

export const testModels = {
	count: {
		// shortName: 'c',
		columns: {
			total: {type: 'INTEGER'},
		},
		migrations: {
			init: {
				up({db, model, queue}) {
					// eslint-disable-next-line jest/no-standalone-expect
					expect(db).toBeTruthy()
					// eslint-disable-next-line jest/no-standalone-expect
					expect(queue).toBeTruthy()
					return model.set({id: 'count', total: 0, byType: {}})
				},
			},
		},
		preprocessor: async ({event}) => {
			if (event.type === 'error_pre') throw new Error('pre error for you')
		},
		reducer: async ({model, event: {type}}) => {
			if (type === 'error_reduce') throw new Error('error for you')
			if (!model.get) return false
			const c = (await model.get('count')) || {
				id: 'count',
				total: 0,
				byType: {},
			}
			c.total++
			c.byType[type] = (c.byType[type] || 0) + 1
			return {
				set: [c],
				// audit: '',
			}
		},
		deriver: async ({event}) => {
			if (event.type === 'error_derive') throw new Error('post error for you')
		},
	},
	ignorer: {
		// eslint-disable-next-line no-unused-vars
		reducer: args => {},
	},
	deriver: {
		deriver: async ({model, store, result, event}) => {
			if (result !== event.result[model.name]) {
				throw new Error('Expecting event.result as separate input')
			}
			if (event.result.count) {
				const currentCount = await store.count.get('count')
				await model.set({
					id: 'descCount',
					desc: `Total: ${currentCount.total}, seen types: ${Object.keys(
						currentCount.byType
					)}`,
				})
			}
		},
	},
}

const withDBs = async fn => {
	const db = new DB({name: 'D'})
	const queue = new EQ({
		db: new DB({name: 'Q'}),
		columns: {events: {type: 'JSON'}},
	})
	const ret = await fn(db, queue)
	await Promise.all([db.close(), queue.db.close()])
	return ret
}

/**
 * @param {Record<string,any>|function} modelsOrFn
 * @param {function} [fn]
 */
export const withESDB = (modelsOrFn, fn) => {
	let models
	if (typeof modelsOrFn === 'function') {
		// eslint-disable-next-line no-throw-literal
		if (fn) throw 'Use either .withESDB(fn) or .withESDB(models, fn)'
		fn = modelsOrFn
	} else {
		models = modelsOrFn
	}
	if (!models) models = testModels
	if (!fn) throw new Error('no fn passed to withESDB')
	return withDBs(async (db, queue) => {
		const eSDB = new ESDB({queue, models, name: 'E'})
		const out = await fn(eSDB, queue)
		await eSDB.close()
		return out
	})
}
