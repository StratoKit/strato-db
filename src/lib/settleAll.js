import {Sema} from 'async-sema'

// Only throw after all items are processed
/**
 * @template Item
 * @param {Item[]} items
 * @param {(item: Item) => Promise<void> | void} fn Function to call on each
 *   item.
 * @param {number} [maxConcurrent] Maximum functions running in parallel.
 */
export const settleAll = async (items, fn, maxConcurrent) => {
	let err, cb
	if (maxConcurrent) {
		const sema = new Sema(maxConcurrent)
		cb = async item => {
			await sema.acquire()
			try {
				return await fn(item)
			} finally {
				sema.release()
			}
		}
	} else {
		cb = fn
	}
	await Promise.all(
		items.map(async i => {
			try {
				await cb(i)
			} catch (error) {
				// last one wins
				err = error
			}
		})
	)
	if (err) throw err
}
