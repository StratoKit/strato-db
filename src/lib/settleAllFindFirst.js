export /**
 *	Execute all fn (asynchronously), if fn returns values, it returns the first result
 *
 * @param {[objects]} items
 * @param {function} fn
 * @returns {any} first positive result
 */
const settleAllFindFirst = async (items, fn) => {
	let err
	const results = await Promise.all(
		items.map(async i => {
			try {
				const res = await fn(i)
				if (res) return res
			} catch (error) {
				// last one wins
				err = error
			}
		})
	)
	if (err) throw err
	if (results?.filter(Boolean).length) return results[0]
	return false
}
