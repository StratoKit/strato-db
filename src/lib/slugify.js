import {deburr} from 'lodash'

const abc = 'abcdefghijklmnopqrstuvwxyz0123456789'
export const randomString = (/** @type {number} */ n) =>
	// eslint-disable-next-line unicorn/no-new-array
	Array.apply(null, new Array(n))
		.map(() => {
			return abc.charAt(Math.floor(Math.random() * abc.length))
		})
		.join('')

export const slugifyString = (name, alwaysResult) => {
	// extract name from i18n objects
	const t =
		typeof name === 'string'
			? name
			: typeof name === 'number'
			? name.toString()
			: name && typeof name === 'object'
			? Object.values(name).find(v => typeof v === 'string' && v)
			: null
	if (!t) {
		if (alwaysResult) return randomString(12)
		throw new Error(`Cannot slugify ${name}`)
	}
	return encodeURIComponent(deburr(t).trim())
		.replaceAll(/(%..|['()_~])/g, '-')
		.replaceAll(/--+/g, '-')
		.toLowerCase()
		.replaceAll(/(^[^\da-z]+|[^\da-z]+$)/g, '')
		.slice(0, 30)
}

// This is not race-safe - only use for write-seldomn things like backoffice or inside transactions
export const uniqueSlugId = async (model, name, colName, currentId) => {
	const slug = slugifyString(name, true)
	let id = slug
	let i = 1
	const where = currentId && {
		[`${model.idColQ} IS NOT ?`]: [currentId],
	}

	while (await model.exists({[colName]: id}, {where})) {
		id = `${slug}-${++i}`
	}
	return id
}
