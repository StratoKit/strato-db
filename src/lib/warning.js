export const DEV = process.env.NODE_ENV !== 'production'
export let deprecated, unknown

if (DEV) {
	const warned = {}
	const warner = type => (tag, msg, conditionFn) => {
		if (warned[tag]) return
		if (conditionFn && !conditionFn()) return
		warned[tag] = true
		// eslint-disable-next-line no-console
		console.warn(new Error(`!!! ${type} ${msg}`))
	}
	deprecated = warner('DEPRECATED')
	unknown = warner('UNKNOWN')
} else {
	deprecated = () => {}
	unknown = () => {}
}
