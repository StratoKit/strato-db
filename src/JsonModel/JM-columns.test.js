import {getModel, sharedSetup} from '../lib/_test-helpers'

const indexesSql = `
	SELECT m.tbl_name || '.' || ifNull(ii.name, m.name) AS col, m.sql
	FROM sqlite_master AS m,
				pragma_index_info(m.name) AS ii
	WHERE m.type='index'
	ORDER BY 1;
`

const withCols = sharedSetup(async () => {
	const m = getModel({
		columns: {
			foo1: {index: 'SPARSE'},
			foo2: {
				type: 'INTEGER',
				value: o => o.foo1 + 1,
				get: false,
				alias: 'foo2',
			},
			foo3: {
				value: o => o.notExists,
				index: 'ALL',
				unique: true,
			},
			fooGet: {real: true, value: () => 3},
		},
	})
	await m.set({id: 'meep', foo1: 5})
	return m
})
test(
	'columns create',
	withCols(async m => {
		expect(m.columnArr).toMatchSnapshot()
		const row = await m.db.get(`SELECT json, foo2 FROM ${m.name}`)
		expect(row).toEqual({json: `{"foo1":5}`, foo2: 6})
	})
)
test(
	'columns order',
	withCols(async m => {
		// id and json are calculated last
		const l = m.columnArr.length
		expect(m.columnArr[l - 2].name).toBe('id')
		expect(m.columnArr[l - 1].name).toBe('json')
	})
)
test(
	'columns get',
	withCols(async m => {
		// columns don't automatically change the original object
		const saved = await m.get('meep')
		expect(saved.foo2).toBeFalsy()
		expect(saved.fooGet).toBe(3)
		saved.id = 'meep2'
		await m.set(saved)
		const row = await m.db.get`SELECT * FROM ${m.name}ID WHERE id = ${saved.id}`
		const json = JSON.parse(row.json)
		// JSON does not include get columns
		expect(json.fooGet).toBeFalsy()
		expect(json.id).toBeFalsy()
		expect(json.foo1).toBe(5)
	})
)
test(
	'columns indexes',
	withCols(async m => {
		// Indexes are created
		const indexes = await m.db.all(indexesSql)
		expect(indexes.some(i => i.col.includes('foo3'))).toBe(true)
		expect(indexes.every(i => !i.col.includes('foo2'))).toBe(true)
	})
)

test('default w/ value()', async () => {
	const m = getModel({columns: {v: {real: true, default: 5}}})
	await m.set({id: 1})
	expect(await m.db.all(`select * from ${m.name}`)).toEqual([
		{id: '1', json: null, v: 5},
	])
	expect(m.columns.v.ignoreNull).toBe(false)
})

test('default w/ sql', async () => {
	const m = getModel({columns: {v: {sql: 'hex(id)', default: 0}}})
	expect(m.makeSelect({attrs: {v: 5}, sort: {v: -1}, cols: ['v']})).toEqual([
		'SELECT ifNull(hex(id),0) AS _0 FROM "testing" tbl WHERE(ifNull(hex(id),0)=?) ORDER BY _0 DESC',
		[5],
		undefined,
		'SELECT COUNT(*) as t from ( SELECT ifNull(hex(id),0) AS _0 FROM "testing" tbl WHERE(ifNull(hex(id),0)=?) )',
		[5],
		false,
	])
	expect(m.columns.v.ignoreNull).toBe(false)
})

test('value w type JSON', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			v: {type: 'JSON'},
		},
	})
	await m.set({v: {whee: true}})
	await m.set({v: 5})
	await m.set({other: true})
	expect(await m.db.all('SELECT * FROM testing')).toEqual([
		{id: 1, json: null, v: '{"whee":true}'},
		{id: 2, json: null, v: 5},
		{id: 3, json: '{"other":true}', v: null},
	])
	expect(await m.all()).toEqual([
		{id: 1, v: {whee: true}},
		{id: 2, v: 5},
		{id: 3, other: true},
	])
})

test('value not real', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			v: {value: o => o.v * 2},
		},
	})
	const result = {id: 1, v: 10}
	expect(await m.set({v: 5})).toEqual(result)
	expect(await m.get(1)).toEqual(result)
})

test('required', async () => {
	const m = getModel({
		columns: {
			foo: {value: o => o.foo, get: true, required: true},
			bar: {
				async value() {
					return this.count()
				},
				get: true,
				required: true,
			},
		},
	})
	expect(m.columns.foo).toHaveProperty('ignoreNull', false)
	await expect(m.set({})).rejects.toThrow('foo')
	await expect(m.set({foo: null})).rejects.toThrow('foo')
	await expect(m.set({foo: 0})).resolves.toHaveProperty('foo', 0)
	await expect(m.set({foo: ''})).resolves.toHaveProperty('foo', '')
	const obj = await m.set({foo: 'hi'})
	expect(obj).toHaveProperty('foo', 'hi')
	expect(obj).toHaveProperty('bar', 2)
})

test('nested JSON', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			c: {type: 'JSON', path: 'a.b.c'},
			a: {type: 'JSON'},
			b: {type: 'JSON', path: 'a.b'},
		},
	})
	await expect(m.set({a: {b: {c: 3}}})).resolves.toEqual({
		id: 1,
		a: {b: {c: 3}},
	})
	await expect(m.get(1)).resolves.toEqual({id: 1, a: {b: {c: 3}}})
	await expect(m.db.get('select * from testing')).resolves.toEqual({
		a: null,
		b: null,
		c: 3,
		id: 1,
		json: null,
	})
})

test('nested JSON null object', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			c: {type: 'JSON', path: 'a.b.c'},
			a: {type: 'JSON'},
			b: {type: 'JSON', path: 'a.b'},
		},
	})
	await expect(m.set({a: {t: 1}})).resolves.toEqual({
		id: 1,
		a: {t: 1, b: {}},
	})
	await expect(m.get(1)).resolves.toEqual({id: 1, a: {t: 1, b: {}}})
	await expect(m.db.get('select * from testing')).resolves.toEqual({
		a: '{"t":1}',
		b: null,
		c: null,
		id: 1,
		json: null,
	})
})

test('JSON alwaysObject', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			a: {type: 'JSON'},
			b: {type: 'JSON', path: 'a.b', alwaysObject: true},
			c: {type: 'JSON', alwaysObject: false},
			d: {path: 'c.d'},
		},
	})
	expect(m.columns.a).toHaveProperty('alwaysObject', true)
	expect(m.columns.c).toHaveProperty('alwaysObject', false)
	await expect(m.set({a: {t: 1}})).resolves.toEqual({
		id: 1,
		a: {t: 1, b: {}},
		c: undefined,
	})
	await expect(m.get(1)).resolves.toEqual({id: 1, a: {t: 1, b: {}}})
	await expect(m.set({c: 1})).resolves.toEqual({
		id: 2,
		a: {b: {}},
		c: 1,
	})
	await expect(m.get(2)).resolves.toEqual({id: 2, a: {b: {}}, c: 1})
})

test('path in json column', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			b: {path: 'a.b'},
			a: {type: 'JSON'},
			c: {real: true},
			d: {path: 'c.d'},
			e: {type: 'JSON', get: false, parse: null},
			f: {path: 'e.f'},
		},
	})
	expect(m.columns.b.jsonCol).toBe('a')
	expect(m.columns.d.jsonCol).toBe('json')
	expect(m.columns.f.jsonCol).toBe('json')
	await m.set({a: {b: 4}})
	await expect(m.searchOne({b: 4})).resolves.toHaveProperty('id', 1)
	await expect(m.get(1)).resolves.toEqual({id: 1, a: {b: 4}})
	await expect(m.db.get('select * from testing')).resolves.toEqual({
		id: 1,
		a: '{"b":4}',
		c: null,
		e: null,
		json: null,
	})
})

test('where/whereVal', () => {
	const m = getModel({
		columns: {
			s: {where: 's<?'},
			f: {where: v => `${v}=?`},
			v: {whereVal: v => [v * 2]},
			w: {where: 'hi', whereVal: () => []},
			o: {whereVal: () => false},
		},
	})
	expect(
		m.makeSelect({attrs: {s: 1, f: 2, v: 3, w: 4, o: 5}}).slice(0, 2)
	).toEqual([
		'SELECT tbl."id" AS _i,tbl."json" AS _j FROM "testing" tbl WHERE(s<?)AND(2=?)AND(json_extract(tbl."json",\'$.v\')=?)AND(hi)',
		[1, 2, 6],
	])
})

test('where(val, origVal)', () => {
	const m = getModel({
		columns: {
			a: {where: (v, o) => `${v}|${o}`, whereVal: v => [v * 2]},
		},
	})
	expect(m.makeSelect({attrs: {a: 1}}).slice(0, 2)).toEqual([
		'SELECT tbl."id" AS _i,tbl."json" AS _j FROM "testing" tbl WHERE(2|1)',
		[2],
	])
})

test('where string', () => {
	expect(() =>
		getModel({
			columns: {
				s: {where: 'noplaceholder'},
			},
		})
	).toThrow('where')
})

test('whereVal truthy not array', () => {
	const m = getModel({
		columns: {
			s: {whereVal: () => true},
		},
	})
	expect(() => m.makeSelect({attrs: {s: 1}})).toThrow('whereVal')
})

test('falsyBool', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			b: {real: true, falsyBool: true},
			c: {real: true, falsyBool: true},
			d: {falsyBool: true},
			e: {falsyBool: true},
			f: {type: 'JSON', falsyBool: true},
			g: {type: 'JSON', falsyBool: true},
		},
	})
	expect(await m.set({b: true, c: false, e: true, f: false, g: true})).toEqual({
		id: 1,
		b: true,
		c: undefined,
		d: undefined,
		e: true,
		f: undefined,
		g: true,
	})
	await expect(m.get(1)).resolves.toEqual({id: 1, b: true, e: true, g: true})
	await expect(m.db.get('select * from testing')).resolves.toEqual({
		id: 1,
		b: 1,
		c: null,
		json: '{"e":true}',
		f: null,
		g: 'true',
	})
})

test('falsyBool query', async () => {
	const m = getModel({
		columns: {
			id: {type: 'INTEGER'},
			b: {real: true, falsyBool: true},
			c: {real: true, falsyBool: true},
			d: {falsyBool: true},
			e: {falsyBool: true},
			f: {type: 'JSON', falsyBool: true},
			g: {type: 'JSON', falsyBool: true},
		},
	})
	await m.set({c: true, d: true, f: true})
	await m.set({b: true, c: false, e: true, f: false, g: true})
	expect(await m.searchOne({c: false})).toHaveProperty('id', 2)
	expect(await m.searchOne({d: ''})).toHaveProperty('id', 2)
	expect(await m.searchOne({f: 0})).toHaveProperty('id', 2)
	expect(await m.searchOne({b: 1})).toHaveProperty('id', 2)
	expect(await m.searchOne({e: 'yes'})).toHaveProperty('id', 2)
	expect(await m.searchOne({g: {}})).toHaveProperty('id', 2)
})

test('columnName: ({columnName}) => columnDefinition', async () => {
	let m
	expect(() => {
		m = getModel({
			columns: {
				foo1: ({columnName}) => ({
					type: 'TEXT',
					value: o => o[columnName] + '!',
				}),
			},
		})
	}).not.toThrow()
	expect(await m.set({foo1: 'hi'})).toHaveProperty('foo1', 'hi!')
	expect(() => {
		getModel({
			columns: {
				foo1: () => ({
					type: 'TEXT',
					badProp: 5,
				}),
			},
		})
	}).toThrow()
})
