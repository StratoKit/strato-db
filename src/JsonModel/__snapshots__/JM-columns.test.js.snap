// Vitest Snapshot v1

exports[`columns create 1`] = `
[
  {
    "alias": "_0",
    "get": false,
    "ignoreNull": true,
    "index": "SPARSE",
    "jsonCol": "json",
    "jsonPath": "foo1",
    "name": "foo1",
    "parts": [
      "foo1",
    ],
    "path": "foo1",
    "quoted": "\\"foo1\\"",
    "select": "json_extract(tbl.\\"json\\",'$.foo1') AS _0",
    "sql": "json_extract(tbl.\\"json\\",'$.foo1')",
    "where": "json_extract(tbl.\\"json\\",'$.foo1')=?",
  },
  {
    "alias": "foo2",
    "get": false,
    "ignoreNull": true,
    "name": "foo2",
    "parts": [
      "foo2",
    ],
    "path": "foo2",
    "quoted": "\\"foo2\\"",
    "real": true,
    "select": "tbl.\\"foo2\\" AS foo2",
    "sql": "tbl.\\"foo2\\"",
    "type": "INTEGER",
    "value": [Function],
    "where": "tbl.\\"foo2\\"=?",
  },
  {
    "alias": "_1",
    "get": false,
    "ignoreNull": false,
    "index": "ALL",
    "jsonCol": "json",
    "jsonPath": "foo3",
    "name": "foo3",
    "parts": [
      "foo3",
    ],
    "path": "foo3",
    "quoted": "\\"foo3\\"",
    "select": "json_extract(tbl.\\"json\\",'$.foo3') AS _1",
    "sql": "json_extract(tbl.\\"json\\",'$.foo3')",
    "unique": true,
    "value": [Function],
    "where": "json_extract(tbl.\\"json\\",'$.foo3')=?",
  },
  {
    "alias": "_2",
    "get": true,
    "ignoreNull": true,
    "name": "fooGet",
    "parts": [
      "fooGet",
    ],
    "path": "fooGet",
    "quoted": "\\"fooGet\\"",
    "real": true,
    "select": "tbl.\\"fooGet\\" AS _2",
    "sql": "tbl.\\"fooGet\\"",
    "type": "BLOB",
    "value": [Function],
    "where": "tbl.\\"fooGet\\"=?",
  },
  {
    "alias": "_i",
    "autoIncrement": undefined,
    "get": true,
    "ignoreNull": false,
    "index": "ALL",
    "name": "id",
    "parts": [
      "id",
    ],
    "path": "id",
    "quoted": "\\"id\\"",
    "real": true,
    "select": "tbl.\\"id\\" AS _i",
    "sql": "tbl.\\"id\\"",
    "type": "TEXT",
    "unique": true,
    "value": [Function],
    "where": "tbl.\\"id\\"=?",
  },
  {
    "alias": "_j",
    "alwaysObject": true,
    "get": true,
    "ignoreNull": true,
    "name": "json",
    "parse": [Function],
    "parts": [],
    "path": "",
    "quoted": "\\"json\\"",
    "real": true,
    "select": "tbl.\\"json\\" AS _j",
    "sql": "tbl.\\"json\\"",
    "stringify": [Function],
    "type": "JSON",
    "where": "tbl.\\"json\\"=?",
  },
]
`;
